src/include/type/transient_value.h

```diff
@@ -0,0 +1,282 @@
+  /**
+   * The copy assignment operator for TransientValue is currently unused, and should be used
+   * sparingly due to its reliance on malloc/free. The assigned to TransientValue will have the same contents as the
+   * other TransientValue, but in its own memory buffers.
+   * @param other TransientValue to copy from
+   * @return reference of object assigned to
+   * @warning No, seriously: don't go crazy with this thing because it will tank performance if the system calls malloc
+   * too much. We saw this with the old Value system in Peloton when Value objects were created and destroyed too
+   * frequently during query execution.
+   */
+  TransientValue &operator=(const TransientValue &other) {
+    if (this != &other) {  // self-assignment check expected
+      if (Type() == TypeId::VARCHAR) {
+        // free VARCHAR buffer
+        delete[] reinterpret_cast<char *const>(data_);
+      }
+      // clear internal buffer
+      data_ = 0;
+      type_ = other.type_;
+      if (Type() != TypeId::VARCHAR) {
+        data_ = other.data_;
+      } else {
+        CopyVarChar(reinterpret_cast<const char *const>(other.data_));
+      }
+    }
+    return *this;
+  }
+
+  /**
+   * Used by TransientValuePeeker to get the internal buffer of the TransientValue. Don't use directly.
+   * @tparam T C type to generate a TransientValue from
+   * @return reinterpreted version of the internal buffer based on the C type template parameter
+   * @see: TransientValuePeeker for actual access to TransientValue's data as a C type
+   */
+  template <typename T>
+  T GetAs() const {
+    return *reinterpret_cast<const T *const>(&data_);
+  }
+
+  /**
+   * Helper method to take a null-terminated C string, allocate a VARCHAR buffer, copy the C string into the buffer, and
+   * then assign the pointer to the TransientValue's internal data buffer.
+   * @param other null-terminated C string to build a VARCHAR buffer for
+   */
+  void CopyVarChar(const char *const c_string) {
+    TERRIER_ASSERT(Type() == TypeId::VARCHAR,
+                   "This TransientValue's type should be set to VARCHAR if this function is being called.");
+    // allocate a VARCHAR buffer
+    const auto *const other_varchar = reinterpret_cast<const char *const>(c_string);
+    const uint32_t length = *reinterpret_cast<const uint32_t *const>(other_varchar);
+    auto *const varchar = new char[length + sizeof(uint32_t)];
+
+    // copy the length field into the VARCHAR buffer
+    *(reinterpret_cast<uint32_t *const>(varchar)) = length;
+
+    // copy the VARCHAR contents into the VARCHAR buffer
+    char *const varchar_contents = varchar + sizeof(uint32_t);
+    const char *const other_varchar_contents = other_varchar + sizeof(uint32_t);
+    std::memcpy(varchar_contents, other_varchar_contents, length);
+    // copy the pointer to the VARCHAR buffer into the internal buffer
+    data_ = reinterpret_cast<uintptr_t>(varchar);
+  }
+
+  TypeId type_ = TypeId::INVALID;
+  uintptr_t data_;
```

Comment:
Not saying we do this in this PR, but this should leave 7 bytes of padding, giving us a total of 15 bytes to actually play with if we ever want to inline small varlens

Reply:
I originally wrote the `data_` field as a byte array and reinterpreted accordingly. I could just change it back now if you think we want those 15 bytes now.

Reply:
I think a TODO would be fine.